dipy/reconst/csdeconv.py

```diff
@@ -397,48 +410,51 @@ def csdeconv(dwsignal, sh_order, X, B_reg, lambda_=1., tau=0.1):
            constrained super-resolved spherical deconvolution.
 
     """
-    # generate initial fODF estimate, truncated at SH order 4
-    fodf_sh = np.linalg.lstsq(X, dwsignal)[0]
-    fodf_sh[15:] = 0
-
-    fodf = np.dot(B_reg, fodf_sh)
+    mu = 1e-5
+    if P is None:
+        P = np.dot(X.T, X)
+    z = np.dot(X.T, dwsignal)
+
+    try:
+        fodf_sh = _solve_cholesky(P, z)
+    except la.LinAlgError:
+        P = P + mu * np.eye(P.shape[0])
+        fodf_sh = _solve_cholesky(P, z)
+    # For the first iteration we use a smooth FOD that only uses SH orders up
+    # to 4 (the first 15 coefficients.
+    fodf = np.dot(B_reg[:, :15], fodf_sh[:15])
     # set threshold on FOD amplitude used to identify 'negative' values
-    threshold = tau * np.mean(np.dot(B_reg, fodf_sh))
+    threshold = tau * np.mean(fodf)
+    k_last = (fodf < threshold)
+
+    # If all fodf values are greater than zero
+    if not k_last.any():
+        return fodf_sh, 0
 
-    k = []
-    convergence = 50
     for num_it in range(1, convergence + 1):
+        # This is the super-resolved trick.  Wherever there is a negative
+        # amplitude value on the fODF, it concatenates a value to the S vector
+        # so that the estimation can focus on trying to eliminate it. In a
+        # sense, this "adds" a measurement, which can help to better estimate
+        # the fodf_sh, even if you have more SH coefficients to estimate than
+        # actual S measurements.
+        H = B_reg[k_last]
+
+        # We use the Cholesky decomposition to solve for the SH coefficients.
+        Q = P + np.dot(H.T, H)
+        fodf_sh = _solve_cholesky(Q, z)
+
+        # Sample the FOD using the regularization sphere and compute k.
         fodf = np.dot(B_reg, fodf_sh)
```

Comment:
A bit unexpected for Q and H to be that slower than fodf. Now that we are on it let's see if this can be improved a bit. @stefanv do you think we can improve the performance of those 3 lines?